lib/active_merchant/billing/gateways/decidir.rb

```diff
@@ -0,0 +1,220 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class DecidirGateway < Gateway
+      self.test_url = 'https://developers.decidir.com/api/v2'
+      self.live_url = 'https://live.decidir.com/api/v2'
+
+      self.supported_countries = ['AR']
+      self.money_format = :cents
+      self.default_currency = 'ARS'
+      self.supported_cardtypes = [:visa, :master, :american_express, :diners_club]
+
+      self.homepage_url = 'http://www.decidir.com'
+      self.display_name = 'Decidir'
+
+      STANDARD_ERROR_CODE_MAPPING = {
+        1 => STANDARD_ERROR_CODE[:call_issuer],
+        2 => STANDARD_ERROR_CODE[:call_issuer],
+        3 => STANDARD_ERROR_CODE[:config_error],
+        4 => STANDARD_ERROR_CODE[:pickup_card],
+        5 => STANDARD_ERROR_CODE[:card_declined],
+        7 => STANDARD_ERROR_CODE[:pickup_card],
+        12 => STANDARD_ERROR_CODE[:processing_error],
+        14 => STANDARD_ERROR_CODE[:invalid_number],
+        28 => STANDARD_ERROR_CODE[:processing_error],
+        38 => STANDARD_ERROR_CODE[:incorrect_pin],
+        39 => STANDARD_ERROR_CODE[:invalid_numbers],
+        43 => STANDARD_ERROR_CODE[:pickup_card],
+        45 => STANDARD_ERROR_CODE[:card_declined],
+        46 => STANDARD_ERROR_CODE[:invalid_number],
+        47 => STANDARD_ERROR_CODE[:card_declined],
+        48 => STANDARD_ERROR_CODE[:card_declined],
+        49 => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        51 => STANDARD_ERROR_CODE[:card_declined],
+        53 => STANDARD_ERROR_CODE[:card_declined],
+        54 => STANDARD_ERROR_CODE[:expired_card],
+        55 => STANDARD_ERROR_CODE[:incorrect_pin],
+        56 => STANDARD_ERROR_CODE[:card_declined],
+        57 => STANDARD_ERROR_CODE[:card_declined],
+        76 => STANDARD_ERROR_CODE[:call_issuer],
+        96 => STANDARD_ERROR_CODE[:processing_error],
+        97 => STANDARD_ERROR_CODE[:processing_error],
+      }
+
+      def initialize(options={})
+        requires!(options, :api_key)
+        super
+      end
+
+      def purchase(money, payment, options={})
+        post = {}
+        add_auth_purchase_params(post, money, payment, options)
+
+        commit(:post, 'payments', post)
+      end
+
+      def authorize(money, payment, options={})
+        post = {}
+        add_auth_purchase_params(post, money, payment, options)
+
+        commit(:post, 'payments', post)
+      end
```

Comment:
Am I missing something or is there nothing distinguishing auth and purchase?

Reply:
Oh they use different creds... hmm.

Reply:
I vaguely recall discussing this, but what are the downsides of having one set of credentials, and distinguishing them by type (`auth_api_key` and `purchase_api_key`)?

Reply:
@curiousepic You are not deceived. The API endpoints for `auth` and `purchase` are identical, as are those for `void` and `refund`. The difference is the `api_key` credential used.

Reply:
The downside rests in how we would determine which credential to use. We could have `auth_api_key` and `purchase_api_key` and require both. We would need to update the transaction methods in AM to pass the `api_key` based on the transaction being performed. I think this is going to be confusing for customers. 

The `api_key` is presented as a single value - at least for the flow we are using. From what I can tell, the `api_key` is based on settings at Decidir.

Reply:
Attaching the email thread for context

> De: Jared Knipp [mailto:jared@spreedly.com] 
> Enviado el: viernes, 7 de junio de 2019 16:46
> Para: Ignacio Castro Ulzurrun <iulzurrun@prismamp.com>
> CC: Lee Jacobs <lee@spreedly.com>; Branden Roden <branden@paymentsfn.com>; Alejandro Garcia <AlGarcia@prismamp.com>; Facundo Islas <FIslas@prismamp.com>
> Asunto: Re: Spreedly - Decidir Integration
> 
> Hi Ignacio, 
> 
> I'm the engineer at Spreedly working on the Decidir gateway integration. 
> 
> Will we receive a sandbox apikey to test all API requests? Or will we need to test with individual apikey's provided in the API documentation for each request type?
> 
> I have a few questions regarding purchase/pre-auth requests that skip tokenization and send payment information inline. The PDF supplement for sending payment info inline with purchase requests does not include an apikey to test with. My attempt to reuse the apikey provided in the API docs for a purchase with a token were unsuccessful. Can you provide an apikey we can use to test this functionality? Is there an account setting required to perform a purchase/pre-auth without first creating a token?
> 
> Thanks,
> 
> Jared
> 
> 
> On Mon, Jun 10, 2019 at 10:23 AM Facundo Islas <FIslas@prismamp.com> wrote:
> 
> Good morning Jared, I give you the PCI key to operate in the test environment:
> 
> [REDACTED API KEY]
> 
> Any questions are available.
> 
> Regards.-
> 
> Facundo Islas
> 
> De: Jared Knipp [mailto:jared@spreedly.com] 
> Enviado el: lunes, 10 de junio de 2019 12:37
> Asunto: Re: Spreedly - Decidir Integration
> 
> Thanks Facundo.
> 
> I'm receiving an error code of `invalid_site`, "datos_site.origin_site_id", when using the key you provided. 
> 
> Is the key you provided the public or private api key? Or is it interchangeable?
> 
> Thanks,
> Jared
> 
> On Mon, Jun 10, 2019 at 2:52 PM Facundo Islas <FIslas@prismamp.com> wrote:
> 
> Jared, in the PCI implementation, a single key is used, since the request is one.
> 
> Could you pass me the request that you are sending?
> 
> Regards.-
> 
> Facundo Islas
> 
> De: Jared Knipp [mailto:jared@spreedly.com] 
> Enviado el: lunes, 10 de junio de 2019 17:13
> Para: Facundo Islas <FIslas@prismamp.com>
> Asunto: Re: Spreedly - Decidir Integration
> 
> Hi Facundo, 
> 
> I realized my mistake. I was including the 'token' field by accident. Once I removed the field, it works.  
> 
> We will need to provide high-level documentation to our customers on how to enable the 'single' key functionality. How does a customer enable this feature? Is there a setting in the admin panel or is it a manual request?
> 
> Thanks,
> Jared
> 
> 
> On Tue, Jun 11, 2019 at 9:29 AM Facundo Islas <FIslas@prismamp.com> wrote:
> 
> Jared good morning, the single key is used in implementations that are PCI clients.
> 
> As such, a special documentation is provided that was provided by Ignacio.
> 
> From Decide we provide the key so that it can operate.
> 
> Regards.-
> 
> Facundo Islas
> 
> De: Jared Knipp [mailto:jared@spreedly.com] 
> Enviado el: martes, 11 de junio de 2019 18:25
> Para: Facundo Islas <FIslas@prismamp.com>
> Asunto: Re: Spreedly - Decidir Integration
> 
> Thanks Facundo.
> 
> Do you support the two-step flow for pre-authorization and capture using the inline payment information and single key approach? If so, can you send me an API key to test with?
> 
> My observation is that the operations below is how we will map to the Decidir API endpoints. I'm unsure how to distinguish a purchase/pre-authorization and void/refund as these use the same endpoints. Is the endpoint functionality determined by the API keys and settings at Decidir?
> 
> purchase  =>  POST /payments
> 
> pre-authorization => POST /payments
> 
> capture (confirmation) => PUT /payments/{payment_id}
> 
> refund => POST /payments/{payment_id}/refunds
> 
> void (of pre-auth) => POST /payments/{payment_id}/refunds
> 
> Thanks,
> 
> Jared
> 
> Facundo Islas <FIslas@prismamp.com>	Wed, Jun 12, 2019 at 7:37 AM
> To: Jared Knipp <jared@spreedly.com>
> Jared good morning, the key to the information provided is to operate in 1 step.
> 
> I pass another apikey to test the entire flow in 2 steps:
> 
> [REDACTED API KEY]
> 
> It is the only key to use all services.
> 
> Any question, notify me.
> 
> Regards.-
> 
> Facundo Islas
>

Reply:
I see; I'm definitely on board with a singular cred since it's tied to a mode. However, I'm wondering if we should distinguish them somehow, such as a gateway level field for which type and throw an error if an unsupported one is requested. In any case, some way to identify an existing gateway as one or the other without having to know which mode a certain api_key is for...

Reply:
🤔It does get weird because the endpoints are the same for `purchase` and `auth`. Nothing prevents a user from making a `purchase` request when they should be doing an `auth`. The request and response look the same to the user.

I'd prefer to add a gateway field for the mode and reserve the using the adapter inheritance model in case we need to support creating payment method tokens prior to purchase/authorize.

Reply:
There is a disclaimer in the [Decidir documentation](https://decidirv2.api-docs.io/1.0/transacciones-en-dos-pasos/flujo-de-una-transaccion-en-dos-pasos) that this must be enabled

> IMPORTANT! 
> Transactions in two steps are included in the special transactions offered by DECIDE. For this reason, in order to enable them, the client must inform  about its use, in order to configure the necessary parameters at the ID Site level. Note that an ID Site that uses this type of Transaction can not implement the classic operation.

Reply:
@jknipp 👍 Sounds good. Let's actually not use the term "mode" since it implies it can be toggled... maybe something like "auth_type"?

Reply:
The `auth_type` is a binary option, so wouldn't it be toggled? Single step (purchase) would be the default (false), auth and capture would need to be enabled (true).

Reply:
That does make sense, I'll leave it to you.